subsys/logging/log_backend_fcb.c

```diff
@@ -0,0 +1,186 @@
+/*
+ * Copyright (c) 2020 Xylem Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <logging/log_backend.h>
+#include <logging/log_core.h>
+#include <logging/log_msg.h>
+#include <logging/log_output.h>
+#include "log_backend_std.h"
+#include <fs/fcb.h>
+#include <assert.h>
+
+static struct flash_sector sectors[CONFIG_LOG_BACKEND_FCB_SECTOR_COUNT] = {
+        { .fs_off = CONFIG_LOG_BACKEND_FCB_SECTOR1_OFFSET,
+          .fs_size = CONFIG_LOG_BACKEND_FCB_SECTOR1_SIZE },
+#if CONFIG_LOG_BACKEND_FCB_SECTOR_COUNT >= 2
+        { .fs_off = CONFIG_LOG_BACKEND_FCB_SECTOR2_OFFSET,
+          .fs_size = CONFIG_LOG_BACKEND_FCB_SECTOR2_SIZE },
+#endif
+#if CONFIG_LOG_BACKEND_FCB_SECTOR_COUNT >= 3
+        { .fs_off = CONFIG_LOG_BACKEND_FCB_SECTOR3_OFFSET,
+          .fs_size = CONFIG_LOG_BACKEND_FCB_SECTOR3_SIZE },
+#endif
+#if CONFIG_LOG_BACKEND_FCB_SECTOR_COUNT >= 4
+        { .fs_off = CONFIG_LOG_BACKEND_FCB_SECTOR4_OFFSET,
+          .fs_size = CONFIG_LOG_BACKEND_FCB_SECTOR4_SIZE },
+#endif
+#if CONFIG_LOG_BACKEND_FCB_SECTOR_COUNT >= 5
+#error "Max 4 log sectors are supported"
+#endif
+};
+
+static uint8_t buf[CONFIG_LOG_BACKEND_FCB_LINE_LEN];
+static struct fcb fcb = {
+        .f_magic = 0xfcb00106, /* "fcb log" */
+        .f_version = 1,
+        .f_sector_cnt = CONFIG_LOG_BACKEND_FCB_SECTOR_COUNT,
+        .f_sectors = sectors,
+};
+
+static int store_line(uint8_t *data, size_t length, void *ctx)
+{
+        struct fcb_entry loc;
+        int rc;
+
+        /* format: [00:00:04.388,000] <inf> this is a log line */
+        /* store only <wrn> and <err> lines. */
+        /* search for < to manage with and without colours */
+        char *p = memchr(data, '<', length);
+        if (!p || (p[1] != 'e' && p[1] != 'w'))
```

Comment:
this is not needed, it should be done on logger level (logger runtime filtering)

Reply:
I'm not so sure that is enough. Filtering on the front-end means all backends get filtered the same, while in this case only the flash backend has the limitations that require filtering.

In my current use case, we want INF and sometimes DEB log entries shown when we there is a a serial port connected and we want WRN and ERR logs saved to flash in case there isn't a serial port connected. There is no way to know if a serial port is connected or not. Filtering in logger means we have to choose either flash-optimised or serial-optimised logging at build time. I don't think this use case is very rare.